staging/src/k8s.io/apiserver/pkg/cel/library/ip.go

```diff
@@ -0,0 +1,307 @@
+//
+// ip
+//
+// Converts a string to an IP address or results in an error if the string is not a valid IP address.
+// The IP address must be an IPv4 or IPv6 address.
+// IPv4-mapped IPv6 addresses (e.g. ::ffff:1.2.3.4) are not allowed.
+// IP addresses with zones (e.g. fe80::1%eth0) are not allowed.
+// Leading zeros in IPv4 address octets are not allowed.
+//
+//	ip(<string>) <IPAddr>
+//
+// Examples:
+//
+//	ip('127.0.0.1') // returns an IPv4 address
+//	ip('::1') // returns an IPv6 address
+//	ip('127.0.0.256') // error
+//	ip(':::1') // error
+//
+// isIP
+//
+// Returns true if a string is a valid IP address.
+// The IP address must be an IPv4 or IPv6 address.
+// IPv4-mapped IPv6 addresses (e.g. ::ffff:1.2.3.4) are not allowed.
+// IP addresses with zones (e.g. fe80::1%eth0) are not allowed.
+// Leading zeros in IPv4 address octets are not allowed.
+//
+//	isIP(<string>) <bool>
+//
+// Examples:
+//
+//	isIP('127.0.0.1') // returns true
+//	isIP('::1') // returns true
+//	isIP('127.0.0.256') // returns false
+//	isIP(':::1') // returns false
+//
+// family / isCanonical / isInCIDR / isUnspecified / isLoopback / isLinkLocalMulticast / isLinkLocalUnicast
+//
+// Returns true if a string is a valid IPv4 address, IPv6 address, or is in its canonical format.
+//
+// - family: returns the IP addresses' family (IPv4 or IPv6) as an integer, either '4' or '6'.
+//
+// - isCanonical: returns true when the IP Address is in its canonical format. There is exactly
+// one canonical form for every IP address, so fields containing IPs in canonical
+// form can just be treated as strings when checking for equality or uniqueness.
+//
+// - isInCIDR: returns true if the IP address is contained in the CIDR.
+// The CIDR must be an IPv4 or IPv6 subnet address with a mask, parsed with cidr.
+//
+// - isUnspecified: returns true if the IP address is the unspecified address.
+// Either the IPv4 address "0.0.0.0" or the IPv6 address "::".
+//
+// - isLoopback: returns true if the IP address is the loopback address.
+// Either an IPv4 address with a value of 127.x.x.x or an IPv6 address with a value of ::1.
+//
+// - isLinkLocalMulticast: returns true if the IP address is a link-local multicast address.
+// Either an IPv4 address with a value of 224.0.0.x or an IPv6 address in the network ff00::/8.
+//
+// - isLinkLocalUnicast: returns true if the IP address is a link-local unicast address.
+// Either an IPv4 address with a value of 169.254.x.x or an IPv6 address in the network fe80::/10.
+//
+// Examples:
+//
+//	ip('127.0.0.1').family() // returns '4''
+//	ip('::1').family() // returns '6'
+//	ip('127.0.0.1').family() == 4 // returns true
+//	ip('::1').family() == 6 // returns true
+//	ip('127.0.0.1').isCanonical() // returns true; all valid IPv4 addresses are canonical
+//	ip('2001:db8::abcd').isCanonical() // returns true
+//	ip('2001:DB8::ABCD').isCanonical() // returns false
```

Comment:
This feels weird to me.  I'd expect the conversion from non-canonical `string` into `ip` to accept the non-canonical input but to convert to some optimal form (e.g. `[2]uint64` in go), meaning it would not retain things like this.  IOW: strings representations of IPs can be canonical or not, but the IP itself is "pure".


Are we saying instead that the internal representation is going to retain the string?

I would have expected this to look more like a free-function `isCanonicalIP(string) bool`

Am I thinking of it wrong?

Reply:
ah, yeah, it would have to be more like `wasCanonicalIP` really...

I had suggested it as a function on the IP rather than as a freestanding string function because that matched the other functions, which I guess matches CEL norms.

Reply:
So we would track a `bool` indicating "we accepted this IP but it was non-canonical" ?  OK I guess.  I'll leave that to @JoelSpeed to decide the right answer

Reply:
The `isCanonical` check as I've implemented is a member of the `IP` type, but it takes the raw/original input string and compares it to the output of the `netip.Addr.String()` which is always canonical. If they are different, it wasn't canonical.

So I think making it freestanding would mean it still has to parse, doing the same work as `ip()` already, so I'm leaning towards keeping it as a member to minimise the top level exposed syntax in the library extension, but also to change the name to `wasCanonicalIP` as suggested by Dan.

Reply:
> but also to change the name to `wasCanonicalIP` as suggested by Dan.

I'm not sure I was actually suggesting that... it feels weird. You are really asking a question about the string ("is this string the canonical representation of some IP address?"), not a question about the IP address that the string represents. So `isCanonical('2001:db8::abcd')` seems like it makes more sense than `ip('2001:db8::abcd').wasCanonical()`.

Reply:
Ok, I'm happy to make it top level and not a member of the `ip()` but I'd like an ack from someone like @cici37 or @jpbetz who has more experience in this than I do

Reply:
Top level functions are safe to add.  CEL [is able to distinguish between](https://playcel.undistro.io/?content=H4sIAAAAAAAAA3WRTUsDQQyG%2F0ocD22hticvvRUpiFQtihRhL%2Bluujswk1nmo7WI%2F93MdrVW6GlCknneN8mnqjCimqlruCdPoAPERh5uU4RckhDe549LcB4eXp%2BfYOu8xTgpuOAqeYza8QwKddsUqmA1VvTRegpB0kKdTmFNpnSWILqOfLdYwsrgofYucXVVsLSc57IHZNGN5LGMekfC2MxDILsxB2jdXoxWQLzT3rEljpktssbJACgAicnrrrLXsTnqOmsdw%2BLXHSyR64Q1wVD0R5PspHNzA2uvI8HBJd9ZO00EjUj3TW%2BBOjB6QpBSjoOuKC%2Foz%2F7GFxbYU1aZDIOXxINuih2ahBfEsUbNIf47UA9a9PPnf6Uzhsp8GXBbIaBtDYXeWGj18Wqn%2Bw1%2FgpH6%2BgakwBpPEQIAAA%3D%3D) function names and variable identifiers.  If you add one, consider prefixing the function name with a scope like how the recently added [`sets.contains`](https://github.com/google/cel-go/blob/master/ext/sets.go) does.